### PR TITLE
Query History: Migrate local storage to remote storage

### LINKED ISSUE #48409: Query History: Migrate local storage to remote storage

### PR DESCRIPTION
**Pull request by ifrost**
**What this PR does / why we need it**:

Opening for earlier review, but it's based on #48330 and requires it to be merged first.

This PR adds automatic migration from local storage query history to the backend database. Your current query history items should be migrated when you open the query history container and an info pop up should be displayed:

Screen Shot 2022-05-02 at 11 20 28

TODO:

- [x] update message "Query history successfully migrated from local storage"

**Which issue(s) this PR fixes**:

Fixes #48409

**Special notes for your reviewer**:

<details><summary>How to test it</summary>

To use a long list of fake query history items please follow the steps below (warning: your current local storage history will be replaced with fake entries).

1. Clean up current entries in db: `sqlite3 data/grafana.db 'delete from query_history; delete from query_history_star'`
2. Get uids and names for some data sources: `sqlite3 data/grafana.db 'select uid,name from data_source where name="gdev-loki" or name="gdev-prometheus"' `

The output should be something like:

```
PDDA8E780A17E7EF1|gdev-loki
PE1C5CBDA0504A6A3|gdev-prometheus
```

Copy that output and paste into JavaScript snippet below:

```javascript
// NOTE: copy paste the output of sqlite3 data/grafana.db 'select uid,name from data_source where name="gdev-loki" or name="gdev-prometheus"'
let datasources = `
PDDA8E780A17E7EF1|gdev-loki
PE1C5CBDA0504A6A3|gdev-prometheus
`.split('\n').filter(l => l).map(l => l.split('|'));

window.localStorage.removeItem('grafana.explore.richHistory.migrated')
window.localStorage.removeItem('grafana.explore.richHistory');

console.log('adding queries...')

let entries = [];
for (let i = 0; i < 10000; i++) {
    let [dsUid, dsName] = datasources[Math.floor(Math.random() * datasources.length)];
    let ts = Math.floor(Date.now() - Math.random() * 14 * 24 * 60 * 60 * 1000);
    let starred = Math.random() > 0.5;
    let query = {
        ts,
        datasourceName: dsName,
        starred,
        comment: `#${i}: starred=${starred}, ts=${ts} (${new Date(ts)}), dsName=${dsName}, dsUid=${dsUid}`,
        queries: [ { expr: `{queryId="${i}"}`, datasource: { uid: dsUid } }]
    }
    entries.push(query);
    try {
        entries.length % 100 === 0 && window.localStorage.setItem('grafana.explore.richHistory', JSON.stringify(entries))    
    } catch (e) {
        console.error(e);
        break;
    }
}

console.log('done');
```

3. Copy paste the snippet into your browser console
4. Wait for the snippet to finish
5. Open query history -> items should be migrated
</details>

### CODE REVIEW

File: public/app/core/utils/richHistory.ts [hunk 1]
```diff
@@ -118,6 +120,27 @@ export async function deleteQueryInRichHistory(id: string) {
...
+      from: 0,
+      search: '',
+      sortOrder: SortOrder.Descending,
+      starred: false,
+      to: 14,
+    });
+    await richHistoryRemoteStorage.migrate(richHistory);
+    return true;
+  } catch (error) {
+    dispatch(notifyApp(createWarningNotification(`Query history migration failed. ${error.message}`)));
```
Comment (gelicia) [inline public/app/core/utils/richHistory.ts:139]: Not sure if this is scope creep, but maybe we would want to track failures in analytics?

Comment (gelicia) [inline public/app/core/utils/richHistory.ts:139]: Not sure if this is scope creep, but maybe we would want to track failures in analytics?

File: public/app/features/explore/state/history.ts [hunk 2]
```diff
@@ -140,6 +145,20 @@ export const clearRichHistoryResults = (exploreId: ExploreId): ThunkResult<void>
  */
 export const initRichHistory = (): ThunkResult<void> => {
   return async (dispatch, getState) => {
+    const queriesMigrated = store.getBool(RICH_HISTORY_SETTING_KEYS.migrated, false);
+    const migrationFailedDuringThisSession = getState().explore.richHistoryMigrationFailed;
+
+    // Query history migration should always be successful, but in case of unexpected errors we ensure
+    // the migration attempt happens only once per session, and the user is informed about the failure
+    // in a way that can help with potential investigation.
+    if (config.queryHistoryEnabled && !queriesMigrated && !migrationFailedDuringThisSession) {
```
Comment (gelicia) [inline public/app/features/explore/state/history.ts:154]: Just wondering if we might want to also check that there was something in local storage to migrate first. Wondering about users who don't keep anything in local storage, can only use incognito tabs or whatever else. It wouldn't be that much overhead to start a migration that didn't have anything, but I think we should be able to check a variable quick here before proceeding

Comment (ifrost) [reply public/app/features/explore/state/history.ts:155]: Good point. In incognito mode it'd make redundant calls to migrate. Updated in e7b5aa3ae8e5b7999a27eccc3d909d56613039bf

**Review by gelicia — APPROVED** [review]

**Pull request merged**